### ISSUE TITLE
network outage, scheduling another check in 60 sec

### ISSUE THREAD

**Issue by pyr0hu** [getsentry/relay#3095]
Hello!

I've installed Relay into our infra to improve the response time for the HTTP API of our application.

We installed relay as the binary, added a static config and started a systemd service which keeps the relay running. After a minute or so, I start to get the network outage in the logs and the relay stops proxying the incoming requests. Enabling debug logs does not show anything out of order. It is deployed in AWS as an EC2 instance. The only third-party stuff on the instance is a Tailscale VPN connection. The sec groups should allow all outgoing connection so I don't think that's the problem as the sentry API calls seems to run OK. (See logs later)

Relay is running using the following systemd unit:

```
[Unit]
Description=Sentry Relay
After=network.target
StartLimitIntervalSec=0

[Service]
Type=simple
Restart=always
RestartSec=3
User=ec2-user
ExecStart=/usr/local/bin/sentry-relay run --config /home/ec2-user/.relay

[Install]
WantedBy=multi-user.target
```

The config folder contains the following:

```
.
├── config.yml
├── credentials.json
└── projects
    └── <project id>.json
```

config.yml

```
relay:
  mode: static
  upstream: https://sentry.io/
  host: 0.0.0.0
  port: 80
logging:
  level: debug
```

Project json is
```
{
  "slug": "<slug>",
  "publicKeys": [
    {
      "publicKey": "<pk>",
      "isEnabled": true
    }
  ],
  "config": {
    "allowedDomains": ["*"]
  }
}
```

Excerpt from the logs:

```
Feb 12 14:52:15 ip-10-0-1-183.eu-west-1.compute.internal sentry-relay[100660]: 2024-02-12T14:52:15.133025Z DEBUG relay_server::services::project_local: loading local states from file system directory="/home/ec2-user/.relay/projects"
Feb 12 14:52:25 ip-10-0-1-183.eu-west-1.compute.internal sentry-relay[100660]: 2024-02-12T14:52:25.132325Z DEBUG relay_server::services::project_local: loading local states from file system directory="/home/ec2-user/.relay/projects"
Feb 12 14:52:29 ip-10-0-1-183.eu-west-1.compute.internal sentry-relay[100660]: 2024-02-12T14:52:29.429381Z  WARN relay_server::services::upstream: network outage, scheduling another check in 60s
Feb 12 14:52:34 ip-10-0-1-183.eu-west-1.compute.internal sentry-relay[100660]: 2024-02-12T14:52:34.360436Z DEBUG request{method=POST uri=/api/<project id>/envelope/ version=HTTP/1.1}: tower_http::trace::on_request: started processing request
Feb 12 14:52:34 ip-10-0-1-183.eu-west-1.compute.internal sentry-relay[100660]: 2024-02-12T14:52:34.362667Z DEBUG request{method=POST uri=/api/<project id>/envelope/ version=HTTP/1.1}: tower_http::trace::on_response: finished processing request latency=2 ms status=200
Feb 12 14:52:35 ip-10-0-1-183.eu-west-1.compute.internal sentry-relay[100660]: 2024-02-12T14:52:35.131801Z DEBUG relay_server::services::project_local: loading local states from file system directory="/home/ec2-user/.relay/projects"
Feb 12 14:52:45 ip-10-0-1-183.eu-west-1.compute.internal sentry-relay[100660]: 2024-02-12T14:52:45.132192Z DEBUG relay_server::services::project_local: loading local states from file system directory="/home/ec2-user/.relay/projects"
Feb 12 14:52:55 ip-10-0-1-183.eu-west-1.compute.internal sentry-relay[100660]: 2024-02-12T14:52:55.132066Z DEBUG relay_server::services::project_local: loading local states from file system directory="/home/ec2-user/.relay/projects"
Feb 12 14:53:05 ip-10-0-1-183.eu-west-1.compute.internal sentry-relay[100660]: 2024-02-12T14:53:05.131871Z DEBUG relay_server::services::project_local: loading local states from file system directory="/home/ec2-user/.relay/projects"
Feb 12 14:53:15 ip-10-0-1-183.eu-west-1.compute.internal sentry-relay[100660]: 2024-02-12T14:53:15.131785Z DEBUG relay_server::services::project_local: loading local states from file system directory="/home/ec2-user/.relay/projects"
Feb 12 14:53:19 ip-10-0-1-183.eu-west-1.compute.internal sentry-relay[100660]: 2024-02-12T14:53:19.447582Z DEBUG request{method=POST uri=/api/<project id>/envelope/ version=HTTP/1.1}: tower_http::trace::on_request: started processing request
Feb 12 14:53:19 ip-10-0-1-183.eu-west-1.compute.internal sentry-relay[100660]: 2024-02-12T14:53:19.449932Z DEBUG request{method=POST uri=/api/<project id>/envelope/ version=HTTP/1.1}: tower_http::trace::on_response: finished processing request latency=2 ms status=200
Feb 12 14:53:21 ip-10-0-1-183.eu-west-1.compute.internal sentry-relay[100660]: 2024-02-12T14:53:21.966891Z DEBUG request{method=POST uri=/api/<project id>/envelope/ version=HTTP/1.1}: tower_http::trace::on_request: started processing request
Feb 12 14:53:21 ip-10-0-1-183.eu-west-1.compute.internal sentry-relay[100660]: 2024-02-12T14:53:21.969276Z DEBUG request{method=POST uri=/api/<project id>/envelope/ version=HTTP/1.1}: tower_http::trace::on_response: finished processing request latency=2 ms status=200
Feb 12 14:53:25 ip-10-0-1-183.eu-west-1.compute.internal sentry-relay[100660]: 2024-02-12T14:53:25.131874Z DEBUG relay_server::services::project_local: loading local states from file system directory="/home/ec2-user/.relay/projects"
```

From the logs, it seems like the envelope calls are returning OK, and there are no indication what network or endpoint has an outage. How can I debug this further?

**Comment by iker-barriocanal** [comment]
Hi!

That's unexpected. Does relay forward events successfully in the first minute? If yes, could there be a service blocking relay for outgoing requests (e.g. the VPN)? Have you tried running the bare relay without additional software (i.e. without the VPN and any other software that may interfere)?

**Comment by pyr0hu** [comment]
@iker-barriocanal Yes, in the first minute I use usages in stats on the web interface, even after the one minute, but only a tenth of what usually arrives at Sentry.io, gets sent from the relay. From what I've understood from the code, the envelope logs are when my application sends the data, and not what relay sends to the upstream. Can I somehow print out or check what does relay want to do on upstream? URL or anything, so I could try to replicate the calls from the actual server with curl.

I've tried disabling the VPN and added a test security group with allow all, but same results.

Additional question: in the config, do I have to set sentry.io as relay.upstream, or the whole upstream from the DNS? o123.ingest.etcetc?

**Comment by pyr0hu** [comment]
Another observation: the events I see on Sentry.io are coming from the JS SDK, not from relay. I started sending events using the sentry-cli from my local computer to relay, and nothing is coming through. It buffers the first events I send, then after the relay wants to send the batch, I got the network outage issue. If I call `curl -X POST <the full DSN>` on the relay server, I get some response back from sentry, so I _can_ connect to Sentry. 

EDIT: 
This part in the docs is a bit confusing for me:

```
In static mode, Relay does not register with upstream since it does not query information from it. After processing events for configured projects, it forwards them upstream with the authentication information (DSN) set by the client that sent the original request.
```

Especially the last part. If I set my DSN to the proxy, how does it use the authentication information from the client? I get it that _key_ is part of the DSN, so the relay can extract it, but the ingest URL is unknown for the relay, isn't it?

**Comment by iker-barriocanal** [comment]
Thanks a lot for the investigation and the details! Some notes answering various questions, let me know if I miss something or if you have further questions:
- Relay does not log the contents of the envelopes but the actions it does, like "finished processing requests".
- In static mode, Relay should only connect with upstream to forward envelopes and doesn't make any other requests.
- Some Relay modes register with upstream because they fetch project configs. In static mode, the project config is hardcoded in the JSON file, so Relay doesn't need to fetch it and thus doesn't need to register with upstream.
- Relay forwards envelopes to the upstream in the config (so in the case above, `sentry.io`), and the upstream will look at the DSN and process the envelope accordingly.
- The only case where a static Relay buffers envelopes is when it can't connect with upstream, to minimize the data loss. If connectivity with upstream is ok, there should be pretty much no buffer in Relay.

Some envelopes may be rejected upstream, but that doesn't explain the unexpected network outage mode. This could be Relay misbehaving or something off in the setup. We'll verify that Relay can forward envelopes in static mode and post updates in this issue.

On the other hand, could you provide some more details on what your use case is? I'd like to learn more about your use case and see if there are any improvements we can make on our end. Thank you beforehand!

**Comment by pyr0hu** [comment]
Thanks @iker-barriocanal for the answers!

My use case is simple, we are running a PHP backend and installed Excimer profiler. Which works using the sentry.io DSN. We wanted to minimize the latency using a relay as per the docs.

Updated notes:

I changed the log level to trace which provided a few additional logs.

After running relay, I sent a message with sentry-cli, got an event UUID back. On the relay I see this:

```
DEBUG request: tower_http::trace::on_request: started processing request method=POST uri=/api/4504020485931008/envelope/ version=HTTP/1.1
DEBUG relay_server::services::project: project <id> state requested 1 times
DEBUG relay_server::services::project: project state <id> updated
DEBUG request: tower_http::trace::on_response: finished processing request latency=0 ms status=200 method=POST uri=/api/<id>/envelope/ version=HTTP/1.1
TRACE relay_server::services::processor: Processing Error group
TRACE relay_server::services::processor::event: processing json event
TRACE relay_server::services::processor: sending envelope to sentry endpoint
```

If I wait and do nothing, only the local project state loader prints outs logs. At this point I'm not sure if my message got through to Sentry, but I don't see anything on the dashboard.

If I send another error, I get this:

```
DEBUG request: tower_http::trace::on_request: started processing request method=POST uri=/api/<id>/envelope/ version=HTTP/1.1
DEBUG request: tower_http::trace::on_response: finished processing request latency=0 ms status=200 method=POST uri=/api/<id>/envelope/ version=HTTP/1.1
TRACE relay_server::services::processor: Processing Error group
TRACE relay_server::services::processor::event: processing json event
TRACE relay_server::services::processor: sending envelope to sentry endpoint
 WARN relay_server::services::upstream: network outage, scheduling another check in 0ns
 WARN relay_server::services::upstream: network outage, scheduling another check in 1s
```

Sending envelope message appear, but without any hesitation, I get the network outage. But it's only coming up after the second send-event CLI call.

If I ran `curl -X POST https://<dsn>@sentry.io/api/<project id>/envelope/`, I get back responses, so I don't think that Sentry is unreachable.

Now, if I stop the server with Ctrl-C, I see these:

```
 INFO relay_server::services::server: Shutting down HTTP server
ERROR relay_server::utils::managed_envelope: dropped envelope: internal error tags.project_key="<project dsn>" tags.has_attachments=false tags.has_sessions=false tags.has_profiles=false tags.has_transactions=false tags.has_replays=false tags.has_checkins=false tags.event_category=Some(Error) cached_summary=EnvelopeSummary { event_category: Some(Error), attachment_quantity: 0, session_quantity: 0, profile_quantity: 0, replay_quantity: 0, checkin_quantity: 0, secondary_transaction_quantity: 0, has_plain_attachments: false, event_metrics_extracted: false, payload_size: 4708 } recomputed_summary=EnvelopeSummary { event_category: Some(Error), attachment_quantity: 0, session_quantity: 0, profile_quantity: 0, replay_quantity: 0, checkin_quantity: 0, secondary_transaction_quantity: 0, has_plain_attachments: false, event_metrics_extracted: false, payload_size: 4708 }
ERROR relay_server::utils::managed_envelope: dropped envelope: internal error tags.project_key="<project dsn>" tags.has_attachments=false tags.has_sessions=false tags.has_profiles=false tags.has_transactions=false tags.has_replays=false tags.has_checkins=false tags.event_category=Some(Error) cached_summary=EnvelopeSummary { event_category: Some(Error), attachment_quantity: 0, session_quantity: 0, profile_quantity: 0, replay_quantity: 0, checkin_quantity: 0, secondary_transaction_quantity: 0, has_plain_attachments: false, event_metrics_extracted: false, payload_size: 4708 } recomputed_summary=EnvelopeSummary { event_category: Some(Error), attachment_quantity: 0, session_quantity: 0, profile_quantity: 0, replay_quantity: 0, checkin_quantity: 0, secondary_transaction_quantity: 0, has_plain_attachments: false, event_metrics_extracted: false, payload_size: 4708 }
TRACE relay_server::services::outcome_aggregator: Outcome aggregation requested: TrackOutcome { timestamp: 2024-02-14T12:44:09.940766896Z, scoping: Scoping { organization_id: 0, project_id: ProjectId(<proejct id>), project_key: ProjectKey("<project dsn>"), key_id: None }, outcome: Invalid(Internal), event_id: Some(EventId(4bee34e6-7d51-41eb-82f4-a5a79349c9dd)), remote_addr: Some(<my ip>), category: Error, quantity: 1 }
ERROR relay_server::utils::managed_envelope: dropped envelope: internal error tags.project_key="<project dsn>" tags.has_attachments=false tags.has_sessions=false tags.has_profiles=false tags.has_transactions=false tags.has_replays=false tags.has_checkins=false tags.event_category=Some(Error) cached_summary=EnvelopeSummary { event_category: Some(Error), attachment_quantity: 0, session_quantity: 0, profile_quantity: 0, replay_quantity: 0, checkin_quantity: 0, secondary_transaction_quantity: 0, has_plain_attachments: false, event_metrics_extracted: false, payload_size: 4708 } recomputed_summary=EnvelopeSummary { event_category: Some(Error), attachment_quantity: 0, session_quantity: 0, profile_quantity: 0, replay_quantity: 0, checkin_quantity: 0, secondary_transaction_quantity: 0, has_plain_attachments: false, event_metrics_extracted: false, payload_size: 4708 }
ERROR relay_server::utils::managed_envelope: dropped envelope: internal error tags.project_key="<project dsn>" tags.has_attachments=false tags.has_sessions=false tags.has_profiles=false tags.has_transactions=false tags.has_replays=false tags.has_checkins=false tags.event_category=Some(Error) cached_summary=EnvelopeSummary { event_category: Some(Error), attachment_quantity: 0, session_quantity: 0, profile_quantity: 0, replay_quantity: 0, checkin_quantity: 0, secondary_transaction_quantity: 0, has_plain_attachments: false, event_metrics_extracted: false, payload_size: 4708 } recomputed_summary=EnvelopeSummary { event_category: Some(Error), attachment_quantity: 0, session_quantity: 0, profile_quantity: 0, replay_quantity: 0, checkin_quantity: 0, secondary_transaction_quantity: 0, has_plain_attachments: false, event_metrics_extracted: false, payload_size: 4708 }
TRACE relay_server::services::outcome_aggregator: Erasing event_id, remote_addr for aggregation: TrackOutcome { timestamp: 2024-02-14T12:44:09.940766896Z, scoping: Scoping { organization_id: 0, project_id: ProjectId(<proejct id>), project_key: ProjectKey("<project dsn>"), key_id: None }, outcome: Invalid(Internal), event_id: Some(EventId(4bee34e6-7d51-41eb-82f4-a5a79349c9dd)), remote_addr: Some(<my ip>), category: Error, quantity: 1 }
TRACE relay_server::services::outcome_aggregator: Outcome aggregation requested: TrackOutcome { timestamp: 2024-02-14T12:43:57.295474699Z, scoping: Scoping { organization_id: 0, project_id: ProjectId(<proejct id>), project_key: ProjectKey("<project dsn>"), key_id: None }, outcome: Invalid(Internal), event_id: Some(EventId(28f274c4-a4d3-4862-8422-e3a62dfe0f4f)), remote_addr: Some(<my ip>), category: Error, quantity: 1 }
TRACE relay_server::services::outcome_aggregator: Erasing event_id, remote_addr for aggregation: TrackOutcome { timestamp: 2024-02-14T12:43:57.295474699Z, scoping: Scoping { organization_id: 0, project_id: ProjectId(<proejct id>), project_key: ProjectKey("<project dsn>"), key_id: None }, outcome: Invalid(Internal), event_id: Some(EventId(28f274c4-a4d3-4862-8422-e3a62dfe0f4f)), remote_addr: Some(<my ip>), category: Error, quantity: 1 }
TRACE relay_server::services::outcome_aggregator: Outcome aggregation requested: TrackOutcome { timestamp: 2024-02-14T12:44:06.286338561Z, scoping: Scoping { organization_id: 0, project_id: ProjectId(<proejct id>), project_key: ProjectKey("<project dsn>"), key_id: None }, outcome: Invalid(Internal), event_id: Some(EventId(88d72304-88c8-4ca4-a8b5-ab4810920e3c)), remote_addr: Some(<my ip>), category: Error, quantity: 1 }
TRACE relay_server::services::outcome_aggregator: Erasing event_id, remote_addr for aggregation: TrackOutcome { timestamp: 2024-02-14T12:44:06.286338561Z, scoping: Scoping { organization_id: 0, project_id: ProjectId(<proejct id>), project_key: ProjectKey("<project dsn>"), key_id: None }, outcome: Invalid(Internal), event_id: Some(EventId(88d72304-88c8-4ca4-a8b5-ab4810920e3c)), remote_addr: Some(<my ip>), category: Error, quantity: 1 }
TRACE relay_server::services::outcome_aggregator: Outcome aggregation requested: TrackOutcome { timestamp: 2024-02-14T12:44:02.062604683Z, scoping: Scoping { organization_id: 0, project_id: ProjectId(<proejct id>), project_key: ProjectKey("<project dsn>"), key_id: None }, outcome: Invalid(Internal), event_id: Some(EventId(caab3a43-25d2-4339-8275-af1cf5e1d5fa)), remote_addr: Some(<my ip>), category: Error, quantity: 1 }
TRACE relay_server::services::outcome_aggregator: Erasing event_id, remote_addr for aggregation: TrackOutcome { timestamp: 2024-02-14T12:44:02.062604683Z, scoping: Scoping { organization_id: 0, project_id: ProjectId(<proejct id>), project_key: ProjectKey("<project dsn>"), key_id: None }, outcome: Invalid(Internal), event_id: Some(EventId(caab3a43-25d2-4339-8275-af1cf5e1d5fa)), remote_addr: Some(<my ip>), category: Error, quantity: 1 }
DEBUG relay_server::utils::garbage: Stop garbage collection thread
 INFO relay_server: relay shutdown complete
```

**Comment by iker-barriocanal** [comment]
I have not been able to reproduce the issue, unfortunately. Could you set up a static relay in your local dev env (with the same configs) and verify whether it also goes into a network outage mode?

**Comment by pyr0hu** [comment]
@iker-barriocanal Good news: I found the issue! To allow binding to port 80, I had to use this:

`sudo setcap CAP_NET_BIND_SERVICE=+eip /usr/local/bin/sentry-relay`

So as you said, I've set up a fresh instance on local, worked okay. Created a new cloud instance, installed relay, still working OK. Installed VPN to make sure the 3rd time that it's not the issue, still worked OK. At this point I was binding to port 3000 for testing. So I knew the only difference was the port. After using setcap, relay stopped working.

So I reinstalled the binary, restarted the systemd service and looks okay so far.

Thanks for the help, I was sure it would be my configuration but nothing pointed to that direction that the setcap would mess with the outgoing requests.
